packages/block-editor/src/components/inserter/hooks/use-patterns-paging.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useState } from '@wordpress/element';
+
+import { useAsyncList, usePrevious } from '@wordpress/compose';
+
+const PAGE_SIZE = 20;
+const INITIAL_INSERTER_RESULTS = 5;
+
+/**
+ * Supplies values needed to page the patterns list client side.
+ *
+ * @param {Array}  currentCategoryPatterns An array of the current patterns to display.
+ * @param {string} currentCategory         The currently selected category.
+ * @param {string} scrollContainerClass    Class of container to scroll when moving between pages.
+ * @param {string} currentFilter           The currently search filter.
+ *
+ * @return {Object} Returns the relevant paging values. (totalItems, categoryPatternsList, numPages, changePage, currentPage)
+ */
+export default function usePatternsPaging(
+	currentCategoryPatterns,
+	currentCategory,
+	scrollContainerClass,
+	currentFilter = ''
+) {
+	const [ currentPage, setCurrentPage ] = useState( 1 );
+	const previousCategory = usePrevious( currentCategory );
+	const previousFilter = usePrevious( currentFilter );
+	if (
+		( previousCategory !== currentCategory ||
+			previousFilter !== currentFilter ) &&
+		currentPage !== 1
+	) {
+		setCurrentPage( 1 );
```

Comment:
Nit: This works but might be harder to understand or maintain. An easier solution would be resetting the state in event handlers instead. See https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers

Reply:
Yeh, agree that would be better, but currently the select category event handler is two parent components up, so seems like it would be a mission and a bit of refactoring involved to achieve this. I suggest we leave it as is for now and look at this when we come to merge the paging components between site editor and post editor - unless you can see a quick way of hooking this into the `onSelectCategory` event handler that I am overlooking.